cpp/splinepy/proximity/proximity.hpp

```diff
@@ -0,0 +1,342 @@
+      for (int i{begin}; i < end; ++i) {
+        grid_points_.IndexToParametricCoordinate(i, parametric_coordinate);
+        coordinates_[i] = spline_(parametric_coordinate);
+      }
+    };
+
+    // n-thread execution
+    splinepy::utils::NThreadExecution(sample_coordinates,
+                                      grid_points_.Size(),
+                                      n_thread);
+
+    // plant a new tree
+    coordinates_cloud_ = std::make_unique<Cloud_>(coordinates_,
+                                                  grid_points_.Size());
+    kdtree_ = std::make_unique<Tree_>(SplineType::kDim, *coordinates_cloud_);
+    kdtree_planted_ = true;
+  }
+
+  /// Make initial guess of choice
+  typename SplineType::ParametricCoordinate_ MakeInitialGuess(
+      InitialGuess initial_guess,
+      const double* goal) const {
+
+
+    if (initial_guess == InitialGuess::MidPoint) {
+      using ReturnValueType =
+          typename SplineType::ParametricCoordinate_::value_type;
+      const auto parametric_bounds =
+          splinepy::splines::GetParametricBounds(spline_);
+
+      // mid point is mean of parametric bounds. doesn't consider the goal.
+      return splinepy::utils::Mean<ReturnValueType>(
+          parametric_bounds[0],
+          parametric_bounds[1]
+      );
+
+    } else if (initial_guess == InitialGuess::KdTree) {
+      if (!kdtree_planted_) {
+        // hate to be aggresive, but here it is.
+        splinepy::utils::PrintAndThrowError(
+            "to use InitialGuess::Kdtree, please first plant the kdtree!"
+        );
+      }
+
+      // good to go. ask the tree
+      const int vote = 1;
+      int id[vote];
+      double distance[vote];
+      kdtree_->knnSearch(goal,
+                         vote /* closest neighbor */,
+                         &id[0],
+                         &distance[0]);
+
+      using ReturnType = typename SplineType::ParametricCoordinate_;
+
+      return 
+          grid_points_.template IndexToParametricCoordinate<ReturnType>(id[0]);
+    } else {
+      //well, shouldn't reach here, but to fight a warning, here it is
+      splinepy::utils::PrintAndThrowError("Invalid initial guess option!");
+      return typename SplineType::ParametricCoordinate_{};
+    }// if initial_guess == ...
+  }
+
+  /*!
+   * Builds RHS and fills "eyeders" matrix, which is also required in LHS.
+   *
+   * "eyeders" refer to collection of spline derivatives where only one
+   * parametric dimension is set to one and everythin else zero, which results
+   * in "eye"-like vectorized derivative query
+   *
+   * @params[in] guess current parametric coordinate guess
+   * @params[in] difference result of `GussMinusQuery()`
+   * @params[out] eyeders
+   * @params[out] rhs
+   */
+  void FillEyeDersAndRhs(
+      const typename SplineType::ParametricCoordinate_& guess,
+      const DArrayD_& difference,
+      PxDMatrixD_& eyeders,
+      PArrayD_& rhs) const {
```

Comment:
Please rename accordingly. eyeders is not a very good description of whats going on here!